### ISSUE TITLE
DoctrineORMQuerySourceIterator is now final

### ISSUE THREAD

**Issue by mpoiriert** [sonata-project/exporter#286]
#### Sonata packages

```
$ composer show --latest 'sonata-project/*'
# Put the result here.

sonata-project/admin-bundle              3.45.1 3.45.1 The missing Symfony Admin Generator
sonata-project/block-bundle              3.14.0 3.14.0 Symfony SonataBlockBundle
sonata-project/cache                     2.0.1  2.0.1  Cache library
sonata-project/core-bundle               3.15.1 3.15.1 Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.4.0  2.4.0  Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.1.5  1.1.5  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.8.1  3.8.1  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.5.0  2.5.0  Symfony SonataEasyExtendsBundle
sonata-project/exporter                  2.0.1  2.0.1  Lightweight Exporter library
sonata-project/formatter-bundle          3.5.0  4.1.2  Symfony SonataFormatterBundle
sonata-project/media-bundle              3.18.1 3.18.1 Symfony SonataMediaBundle
sonata-project/user-bundle               4.2.3  4.2.3  Symfony SonataUserBundle

```

#### Symfony packages

```
$ composer show --latest 'symfony/*'
# Put the result here.
symfony/dependency-injection v4.2.2             v4.2.2             Symfony DependencyInjection Component
symfony/doctrine-bridge      v4.2.2             v4.2.2             Symfony Doctrine Bridge
symfony/dom-crawler          v4.2.2             v4.2.2             Symfony DomCrawler Component
symfony/dotenv               v4.2.2             v4.2.2             Registers environment variables from a .env file
symfony/event-dispatcher     v4.2.2             v4.2.2             Symfony EventDispatcher Component
symfony/expression-language  v4.2.2             v4.2.2             Symfony ExpressionLanguage Component
symfony/filesystem           v4.2.2             v4.2.2             Symfony Filesystem Component
symfony/finder               v4.2.2             v4.2.2             Symfony Finder Component
symfony/flex                 v1.1.8             v1.1.8             Composer plugin for Symfony
symfony/form                 v4.2.2             v4.2.2             Symfony Form Component
symfony/framework-bundle     v4.2.2             v4.2.2             Symfony FrameworkBundle
symfony/http-foundation      v4.2.2             v4.2.2             Symfony HttpFoundation Component
symfony/http-kernel          v4.2.2             v4.2.2             Symfony HttpKernel Component
symfony/inflector            v4.2.2             v4.2.2             Symfony Inflector Component
symfony/intl                 v4.2.2             v4.2.2             A PHP replacement layer for the C intl extension that includes additional data from the ICU library.
symfony/lock                 v4.2.2             v4.2.2             Symfony Lock Component
symfony/lts                  dev-master c1affae dev-master c1affae Enforces Long Term Supported versions of Symfony components
symfony/maker-bundle         v1.11.3            v1.11.3            Symfony Maker helps you create empty commands, controllers, form classes, tests and more so you can forge...
symfony/monolog-bridge       v4.2.2             v4.2.2             Symfony Monolog Bridge
symfony/monolog-bundle       v3.3.1             v3.3.1             Symfony MonologBundle
symfony/options-resolver     v4.2.2             v4.2.2             Symfony OptionsResolver Component
symfony/orm-pack             v1.0.6             v1.0.6             A pack for the Doctrine ORM
symfony/panther              v0.2.0             v0.2.0             A browser testing and web scraping library for PHP and Symfony.
symfony/phpunit-bridge       v4.2.2             v4.2.2             Symfony PHPUnit Bridge
symfony/polyfill-ctype       v1.10.0            v1.10.0            Symfony polyfill for ctype functions
symfony/polyfill-intl-icu    v1.10.0            v1.10.0            Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring    v1.10.0            v1.10.0            Symfony polyfill for the Mbstring extension
symfony/polyfill-php72       v1.10.0            v1.10.0            Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/process              v4.2.2             v4.2.2             Symfony Process Component
symfony/profiler-pack        v1.0.4             v1.0.4             A pack for the Symfony web profiler
symfony/property-access      v4.2.2             v4.2.2             Symfony PropertyAccess Component
symfony/routing              v4.2.2             v4.2.2             Symfony Routing Component
symfony/security-acl         v3.0.1             v3.0.1             Symfony Security Component - ACL (Access Control List)
symfony/security-bundle      v4.2.2             v4.2.2             Symfony SecurityBundle
symfony/security-core        v4.2.2             v4.2.2             Symfony Security Component - Core Library
symfony/security-csrf        v4.2.2             v4.2.2             Symfony Security Component - CSRF Library
symfony/security-guard       v4.2.2             v4.2.2             Symfony Security Component - Guard
symfony/security-http        v4.2.2             v4.2.2             Symfony Security Component - HTTP Integration
symfony/stopwatch            v4.2.2             v4.2.2             Symfony Stopwatch Component
symfony/swiftmailer-bundle   v3.2.5             v3.2.5             Symfony SwiftmailerBundle
symfony/templating           v4.2.2             v4.2.2             Symfony Templating Component
symfony/test-pack            v1.0.5             v1.0.5             A pack for functional and end-to-end testing within a Symfony app
symfony/translation          v4.2.2             v4.2.2             Symfony Translation Component
symfony/twig-bridge          v4.2.2             v4.2.2             Symfony Twig Bridge
symfony/twig-bundle          v4.2.2             v4.2.2             Symfony TwigBundle
symfony/validator            v4.2.2             v4.2.2             Symfony Validator Component
symfony/var-dumper           v4.2.2             v4.2.2             Symfony mechanism for exploring and dumping PHP variables
symfony/var-exporter         v4.2.2             v4.2.2             A blend of var_export() + serialize() to turn any serializable data structure to plain PHP code
symfony/web-profiler-bundle  v4.2.2             v4.2.2             Symfony WebProfilerBundle
symfony/yaml                 v4.2.2             v4.2.2             Symfony Yaml Component

```

#### PHP version

```
$ php -v
# Put the result here.
PHP 7.2.3 (cli) (built: Mar 22 2018 22:03:09) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Xdebug v2.6.1, Copyright (c) 2002-2018, by Derick Rethans
```

## Subject

I just updated to the latest Sonata\Export version and the  Sonata\Exporter\Source\DoctrineORMQuerySourceIterator has all is method now final. I don't why this one specifically since all the other SourceIterator are not final.
I was extending it to support Doctrine\ORM\Query::HYDRATE_SCALAR but now I need to copy paste all the content of the class and do some adjustment instead of overiding 2 methods.

**Comment by mpoiriert** [comment]
Sorry about the readme the update of sonata-project/export was made trough a dependency that I didn't update to a major version so I didn't expect to see breaking change.

I just find it odd that all the other SourceIterator are not final but this one is...

The use case is that use a doctrine QueryBuilder  with a custom select and iterate with Query::HYDRATE_SCALAR to generate a csv that would be too complex (probably impossible) with the other hydrate, not taking in consideration the load it save on the system.

For the PR itself I just read your guide line and since it doesn't seem to have any virtualisation (aka Docker) to work it would be hard to follow your requirement. I do not install php or python on my computer. I do follow Symfony coding standard trough my IDE and I can code the test in a existing project I have running and copy paste them when it's done and rely on travis for the "final" execution. But I think at the size of Sonata Project a normalise setup should be provided to help contribution.

I could provide the first iteration of the PR really quickly just to provide a "what" it does and will go from there ?

**Comment by core23** [comment]
> Sorry about the readme the update of sonata-project/export was made trough a dependency that I didn't update to a major version so I didn't expect to see breaking chang

That's why you should put all dependencies in your local `composer.json` file, which you use for your application.

**Comment by OskarStark** [comment]
It should work if you require exporter ^1.0 directly in your composer.json

**Comment by greg0ire** [comment]
@mpoiriert you can use https://github.com/maglnet/ComposerRequireChecker to detect dependencies you don't know you have that shoud appear in your composer.json.

> The use case is that use a doctrine QueryBuilder with a custom select and iterate with Query::HYDRATE_SCALAR to generate a csv that would be too complex (probably impossible) with the other hydrate, not taking in consideration the load it save on the system.

Can you show us some code?

**Comment by mpoiriert** [comment]
Thanks for the package I didn't know about this one !

Yes it's pretty much 2 line of code that need to change to support that. A bit more to make the class support both (argument in constructor, if, etc...) but this is the explanation of the 2 methods I modified as of now:


```PHP
public function current()
{
    $current = $this->iterator->current();

    $data = [];

    foreach ($this->propertyPaths as $name => $propertyPath) {
        try {
            $data[$name] = $this->getValue($this->propertyAccessor->getValue($current[0], $propertyPath));
        } catch (UnexpectedTypeException $e) {
            //non existent object in path will be ignored
            $data[$name] = null;
        }
    }
    //No object in scalar so this line is removed
    //$this->query->getEntityManager()->getUnitOfWork()->detach($current[0]);
    return $data;
}
```

```PHP
public function rewind(): void
{
    if ($this->iterator) {
        throw new InvalidMethodCallException('Cannot rewind a Doctrine\ORM\Query');
    }
    // Iterate specifying scalar
    $this->iterator = $this->query->iterate(null, Query::HYDRATE_SCALAR);
    $this->iterator->rewind();
}
```

**Comment by greg0ire** [comment]
> Yes it's pretty much 2 line of code that need to change to support that.

Please please please make a PR. If you don't want to, you will not be stuck though: you can still use decoration instead of inheritance to achieve your goal, but please make a PR and let everyone benefit from your changes.

As for the docker idea, it's a good idea but it should not be mandatory, contributions would be welcome but this one will be harder.

**Comment by mpoiriert** [comment]
I'll try to do this in the next couple of days.

**Comment by alinmigea** [comment]
Hi @mpoiriert, is there any updates regarding this issue? It's been a blocker for me also and I look forward for an update.
